File: BaseClasses/BaseClasses/Services/ApplicationConfigManager.m
//
//  ApplicationConfigManager.m
//  Life
//
//  Created by William Hindenburg on 1/3/15.
//  Copyright (c) 2015 William Hindenburg. All rights reserved.
//

#import "ApplicationConfigManager.h"
#import "ConfigService.h"

@interface ApplicationConfigManager()
@property (strong, nonatomic) Config *config;
@property (strong, nonatomic) NSTimer *timer;
@end

@implementation ApplicationConfigManager

+ (instancetype)sharedManager {
    static ApplicationConfigManager *sharedManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedManager = [[self alloc] initPrivate];
    });
    return sharedManager;
}

- (id)initPrivate {
    self = [super init];
    if (self) {
        
    }
    return self;
}

- (void)fetchConfig {
    ConfigService *configService = [ConfigService new];
    [configService loadConfigDataWithSuccessBlock:^(Config *config) {
        [self.timer invalidate];
        NSNumber *timeInterval = config.configRefreshIntervalSeconds;
        self.timer = [NSTimer scheduledTimerWithTimeInterval:[timeInterval integerValue] target:self selector:@selector(fetchConfig) userInfo:nil repeats:NO];
        self.config = config;
    } andError:^(NSError *error) {
        NSLog(@"%@", error);
    }];
}

- (BOOL)isInMaintenanceMode {
    NSNumber *isOnNumber = [[self maintenanceModeDictionary] objectForKey:@"isOn"];
    BOOL isOn = [isOnNumber boolValue];
    return isOn;
}

- (NSString *)maintenanceModeMessage {
    NSString *maintenanceModeString = [[self maintenanceModeDictionary] objectForKey:@"message"];
    return maintenanceModeString;
}

- (BOOL)shouldForceUpgrade {
    NSDictionary *forceUpgradeDictionary = [self forceUpgradeDictionary];
    NSNumber *shouldForceUpgrade = [forceUpgradeDictionary valueForKey:@"forceUpgrade"];
    return [shouldForceUpgrade boolValue];
}

- (NSString *)forceUpgradeMessage {
    NSDictionary *forceUpgradeDictionary = [self forceUpgradeDictionary];
    return [forceUpgradeDictionary valueForKey:@"forceUpgradeMessage"];
}

- (NSString *)forceUpgradeTitle {
    NSDictionary *forceUpgradeDictionary = [self forceUpgradeDictionary];
    return [forceUpgradeDictionary valueForKey:@"forceUpgradeTitle"];
}

- (NSString *)forceUpgradeURL {
    NSDictionary *forceUpgradeDictionary = [self forceUpgradeDictionary];
    return [forceUpgradeDictionary valueForKey:@"forceUpgradeURL"];
}

- (NSDictionary *)forceUpgradeDictionary {
    NSDictionary *infoDictionary = [[NSBundle mainBundle] infoDictionary];
    NSString *buildNumber = [infoDictionary objectForKey:(NSString *)kCFBundleVersionKey];
    NSString *versionNumber = [[infoDictionary objectForKey:@"CFBundleShortVersionString"] stringByReplacingOccurrencesOfString:@"." withString:@""];
    
    NSDictionary *forceUpgradeDictionary = [self topLevelForceUpgradeDictionary];
    NSDictionary *versionDictionary = [forceUpgradeDictionary objectForKey:versionNumber];
    NSDictionary *buildDictionary = [versionDictionary objectForKey:buildNumber];
    return buildDictionary;
}

- (NSDictionary *)maintenanceModeDictionary {
    return self.config.MaintenanceMode;
}

- (NSDictionary *)topLevelForceUpgradeDictionary {
    return self.config.forceUpgrade;
}

@end

File: BaseClasses/BaseClasses/Services/ApplicationConfigManager.swift
//
//  ApplicationConfigManager.swift
//  BaseClasses
//
//  Created by William Hindenburg on 7/19/15.
//  Copyright © 2015 William Hindenburg. All rights reserved.
//

import Foundation

@objc public class ApplicationConfigManager: NSObject {
    public static let sharedManager = ApplicationConfigManager()
    private var config: Config?
    private var timer = NSTimer()
    
    public func isInMaintenanceMode() -> Bool {
        if let isOn = maintenanceModeDictionary()["isOn"] as? NSNumber {
            return isOn.boolValue
        } else {
            return false
        }
        
    }
    
    public func maintenanceModeMessage() -> String {
        return maintenanceModeDictionary()["message"] as! String
    }
    
    public func shouldForceUpgrade() -> Bool {
        if let shouldForceUpgrade = self.forceUpgradeDictionary()?["forceUpgrade"] as? NSNumber {
            return shouldForceUpgrade.boolValue
        } else {
            return false
        }
        
        
    }
    
    public func forceUpgradeMessage() -> String {
        return forceUpgradeDictionary()?["forceUpgradeMessage"] as? String ?? ""
    }
    
    public func forceUpgradeTitle() -> String {
        return forceUpgradeDictionary()?["forceUpgradeTitle"] as? String ?? ""
    }
    
    public func forceUpgradeURL() -> String {
        return forceUpgradeDictionary()?["forceUpgradeURL"] as? String ?? ""
    }
    
    public func fetchConfig () {
        let configService = ConfigService()
        configService.loadConfigDataWithSuccessBlock({ (newConfig) -> Void in
            self.timer.invalidate()
            self.config = newConfig
            if (self.config != nil && self.config!.configRefreshIntervalSeconds != nil) {
                let number: NSTimeInterval = self.config!.configRefreshIntervalSeconds!.doubleValue
                            self.timer = NSTimer.scheduledTimerWithTimeInterval(number, target: self, selector: "fetchConfig", userInfo: nil, repeats: false)
                
            }
            }) { (error) -> Void in
                
        }
    }
    
    private func forceUpgradeDictionary() -> [String: AnyObject]? {
        var buildDictionary = [String: AnyObject]?()
        if (NSBundle.mainBundle().infoDictionary != nil) {
            let infoDictionary = NSBundle.mainBundle().infoDictionary!
            let buildNumber = infoDictionary[kCFBundleVersionKey as String] as? String
            print(buildNumber)
            var versionNumber = infoDictionary["CFBundleShortVersionString"] as? String
            versionNumber = versionNumber?.stringByReplacingOccurrencesOfString(".", withString: "")
            print(versionNumber)
            if (self.config != nil && self.config!.forceUpgrade != nil && versionNumber != nil && buildNumber != nil) {
                let forceUpgradeDictionary = self.config!.forceUpgrade! as [String: AnyObject]
                print(forceUpgradeDictionary)
                let versionDictionary = forceUpgradeDictionary[versionNumber!] as! [String: AnyObject]
                print(versionDictionary)
                buildDictionary = versionDictionary[buildNumber!] as? [String: AnyObject]
                return buildDictionary
            } else {
                return buildDictionary
            }
        } else {
            return buildDictionary
        }
    }
    
    private func maintenanceModeDictionary() -> [NSObject: AnyObject] {
        return config?.MaintenanceMode ?? [NSObject: AnyObject]()
    }
}


